### ISSUE TITLE
GenericValue operator[] has unnecessary assert inside?

### ISSUE THREAD

**Issue by rcdailey** [miloyip/rapidjson#653]
I upgrade rapidjson from a pretty old version and I am asserting like crazy in this function:

```
    //! Get a value from an object associated with the name.
    /*! \pre IsObject() == true
        \tparam SourceAllocator Allocator of the \c name value

        \note Compared to \ref operator[](T*), this version is faster because it does not need a StrLen().
        And it can also handle strings with embedded null characters.

        \note Linear time complexity.
    */
    template <typename SourceAllocator>
    GenericValue& operator[](const GenericValue<Encoding, SourceAllocator>& name) {
        MemberIterator member = FindMember(name);
        if (member != MemberEnd())
            return member->value;
        else {
            RAPIDJSON_ASSERT(false);    // see above note
            static GenericValue NullValue;
            return NullValue;
        }
    }
```

Notice the `RAPIDJSON_ASSERT(false)` above. I do not understand why this assertion is there, especially since in release builds it has a perfectly reasonable default behavior (return a blank value). The comment block also does not indicate a precondition that the element I'm searching for must exist.

Seems like this assertion should not exist. Why is it there? What should I be doing instead?

**Comment by pah** [comment]
The main reason is that there is no possible protection against *changes* to the returned `NullValue` object, which can lead to surprising bugs.

See [Issue 7](https://code.google.com/archive/p/rapidjson/issues/7) in the old GoogleCode project:
````cpp
Value o(kObjectType); o["foo"] = "foo"; o["bar"] = "bar";
RAPIDJSON_ASSERT( o["foo"] != o["bar"] ); // BOOM!
````

**Comment by rcdailey** [comment]
Can you put the assert in a dedicated non const overload then? Or provide a real getter function that doesn't allow modification of the returned lvalue?

**Comment by pah** [comment]
Adding a separate `const` overload returning a `const` reference instead of asserting could be an option.  Can you elaborate on the possible implementation of the second idea?

**Comment by rcdailey** [comment]
Sorry for the late response.

The problem is convention. Most people will assume `operator[]` always returns a mutable reference and in addition adds the element if it does not exist. If you do not want these semantics, I would say the best option is to get rid of `operator[]` entirely. So roughly:

```
template <typename SourceAllocator>
GenericValue& operator[](const GenericValue<Encoding, SourceAllocator>& name) {
    MemberIterator member = FindMember(name);
    if (member != MemberEnd())
        return member->value;
    else {
        // Create element, with empty value
        return new_member->value;
    }
}
```

For scenarios where you want to obtain the value of a key *if it exists already* or return a blank value, you can add a `GenericValue const& get() const;` function, like so:

```
template <typename SourceAllocator>
GenericValue& get(const GenericValue<Encoding, SourceAllocator>& name) {
    MemberIterator member = FindMember(name);
    if (member != MemberEnd())
        return member->value;
    else {
        static GenericValue const dummy_value;
        return dummy_value->value;
    }
}
```

This has the benefit of being unconventional (i.e. there is no `get()` in things like `std::map`, so there will be no misunderstandings on semantic behavior). It also provides you the guarantee that you can use a static variable without concerns over mutability.

I'm being rather high level here but hopefully that makes sense. I have written classes that provide access to application settings (simply key/value pairs) before, and I used both `operator[]` (to provide semantics similar to `std::map`, so you could use the settings object as a drop in replacement for a map) and `get()` (to provide expanded functionality on top of what is expected from `std::map` to handle mutability concerns).

Hopefully this helps give you some ideas.

**Comment by pah** [comment]
> The problem is convention. Most people will assume operator[] always returns a mutable reference and in addition adds the element if it does not exist. If you do not want these semantics, I would say the best option is to get rid of operator[] entirely. 

I agree, that this is an inconvenient deviation from the C++ standard library.  Unfortunately, the RapidJSON allocator API currently doesn't allow a transparent creation of the "missing" element as the `GenericValue` has no access to a matching allocator.

OTOH, removing the `operator[]` entirely would break a lot of existing code.

**Comment by isaacselement** [comment]
rapidjson::Value &config = .... ;  // already initialize .
So

    rapidjson::Value& sizeVal = config["Key"];
    if (!sizeVal.IsNull()) {
         // do something with sizeVal ...
    }

Codes now will crash , because of `RAPIDJSON_ASSERT(false); ` in `operator[]` .  Old version is fine . Then I have to refactor the codes: 

    if (config.HasMember("Key")) {
        rapidjson::Value& sizeVal = config["Key"];
        // do something with sizeVal ...
    }

I don't think that put this `RAPIDJSON_ASSERT(false)` is a good design or solution . Cause it make everyone now have to write the Key/Name twice .

**Comment by fboranek** [comment]
I think the assert is essential here. I just refactored code and switch to RapidJSON. The old code looks like:

```root["result"]["code"] = code;```

which doesn't work with RapidJSON at all. At least the assert is there. I would prefer not to compile.
